### ISSUE TITLE
Allow e2 coroutines to modify values outside of their thread

### ISSUE THREAD

**Issue by Vurv78** [Vurv78/VExtensions#18]
Forgot to make this an issue a while ago.
Since we end up copying the scope frozen in time, it can't modify the global scope that currently occurs when it's resumed, so maybe set it's __newindex to effect both the frozen and current scope?

**Comment by Cheatoid** [comment]
Well, it is not really a bug, it is just how threads (coroutines) work. This is to be addressed within the Second Patch (already listed on the issue page: #17):
> Add vararg support for `try` & `coroutine:resume`/`yield`

-----

When you initially resume the unstarted coroutine, the coroutine creates its own (private) copy of memory that contains state captured at the creation of a coroutine.
To demonstrate what is going on, run the following E2, for example:
```golo
@outputs Out Data:table
@persist Foo:string

Out = 123
Data = table((1) = "Main stuff")
Foo = "Hello World"
Who = entity() # Temporary variable

function anotherThread()
{
    print( format( "[another thread] Out: %s", Out ) )
    print( format( "[another thread] Data[1]: %s", Data[1, string] ) )
    print( format( "[another thread] Foo: %q", Foo ) )
    print( format( "[another thread] Who: %s", Who ) )
    Out = 9001
    Data[1, string] = "Another thread stuff"
    Foo = "Another thread says hi"
    Who = owner()
}

print( format( "[main thread][before] Out: %s", Out ) )
print( format( "[main thread][before] Data[1]: %s", Data[1, string] ) )
print( format( "[main thread][before] Foo: %q", Foo ) )
print( format( "[main thread][before] Who: %s", Who ) )

coroutine( "anotherThread" ):resume()

print( format( "[main thread][after] Out: %s", Out ) )
print( format( "[main thread][after] Data[1]: %s", Data[1, string] ) )
print( format( "[main thread][after] Foo: %q", Foo ) )
print( format( "[main thread][after] Who: %s", Who ) )
```

Any changes you make to any kind of variables (`@persist` / `@outputs` / temporary) all of these live in TLS (technically called "Thread Local Storage") within the coroutine's private memory space, so all the changes you make within another thread are **not** visible to the main thread (unless transferred via `yield`), and also the memory of the coroutine is gone when the coroutine itself dies and the context is switched back to the main thread.
Therefore the output of the above E2 would be (similar to this):
```
[main thread][before] Out: 123
[main thread][before] Data[1]: Main stuff
[main thread][before] Foo: "Hello World"
[main thread][before] Who: Entity [68][gmod_wire_expression2]
[another thread] Out: 123
[another thread] Data[1]: Main stuff
[another thread] Foo: "Hello World"
[another thread] Who: Entity [68][gmod_wire_expression2]
[main thread][after] Out: 123
[main thread][after] Data[1]: Main stuff
[main thread][after] Foo: "Hello World"
[main thread][after] Who: Entity [68][gmod_wire_expression2]
```
Notice the output is the same for all. The main thread isn't seeing the changes made in another thread.
The only way for the main thread to see/get the changes from an another thread is to use [`yield`](https://wiki.facepunch.com/gmod/coroutine.yield) function within the coroutine function; [`resume`](https://wiki.facepunch.com/gmod/coroutine.resume) does the opposite, passing memory/data from the main to another thread.
Basically, you just need to transfer memory/state between threads; [`yield`](https://wiki.facepunch.com/gmod/coroutine.yield)/[`resume`](https://wiki.facepunch.com/gmod/coroutine.resume) functions does that (assuming you take advantage of the vararg parameter of course).

**Comment by Vurv78** [comment]
Yes, I do know how coroutines work considering I spent several months trying to get this to work with e2..

```lua
-- sf example of what I mean
local coolvar = 100
local thread = coroutine.create(function()
    coolvar = 200
    print("hello!")
    coroutine.yield()
    coolvar = 500
    coroutine.yield()
end)

coroutine.resume(thread)
-- > hello!
print(coolvar) --> 200
```
This is what I mean. The coroutine won't be able to do anything to the persisted values in e2, because there's no support for that. You can see if you pass (self) instead of a cloned self, it will actually affect the main table even if it is in another thread

(We copy it because if that happened, the scopes would be changing and a lot would break)

While if we did it in the current state of coroutine-core..
```golo
@name Coroutine No-Access to main thread vars
@persist Coolvar

Coolvar = 100
function thread(){
    Coolvar = 200
    print("hello!")
    coroutineYield()
    Coolvar = 500
    coroutineYield()
}
Thread = coroutine("thread")

Thread:resume()
print(Coolvar) # --> 100
```

I think this would lead to a ton of confusion

**Comment by Vurv78** [comment]
But yeah, It's not really a bug, just not implemented

**Comment by Cheatoid** [comment]
Right. I been thinking about a way to approach it, there is a problem in that E2 can't return vararg (many return values) like Lua is able to. Which means we can't define such:
`e2function ... coroutine:resume`
The best solution I could think of so far, is to have two sets of overloads for the `resume`/`yield`.
- [x] One set of overloads which doesn't care about transferring state. Equivalent to currently implemented functions, so this is already completed.
- [ ] Instead of doing vararg support (which would require `luaTableToE2` to work perfectly), we could just have overloads that are able to pass state using E2 `table`, since E2 `table` is the most powerful data-type (it can hold nested tables, arrays, any other values except `xgt` global-table but that one is irrelevant).
So, we could define these:
- Yield set of overload methods:
    - `e2function void coroutine:yield()` - the current method, does *not* allow for data transfer
    - `e2function table coroutine:yield(table data)` - a new overload, allowing to exchange state back to the main thread
- Resume set of overload methods:
    - `e2function void coroutine:resume()` - the current method, does *not* allow for data transfer
    - `e2function table coroutine:resume(table data)` - a new overload, allowing to exchange state to the coroutine thread

And be done with it. The easy way out, I can do these overloads right away with PR 😄
What do you think? Should we just go with `table` instead of having to mess with `luaTableToE2`+`buildBody` stuff for vararg support?

**Comment by Cheatoid** [comment]
The same trick can be applied to the `try` e2function.

**Comment by Vurv78** [comment]
Trying to solve this is so annoying, it'd require a rewrite and this is as far as I've gotten right now.
```lua
local function parentTable(t,parent)
    local stockpile = table_copy(parent)
    local meta = {
        __index = function(t,k)
            if stockpile[k] then return stockpile[k] end
            return parent[k]
        end,
        __newindex = function(t,k,v)
            if stockpile[k] then stockpile[k] = v end
            if parent[k] then parent[k] = v end
        end
    }
    return setmetatable(t,meta)
end
```
Ofc it does allow for modifying the scope properly, but if it runs over several ticks it forgets the scope.. I think it's because we'd have to make the metatable set itself recursively on the e2 scope instance or whatever. Really annoying.
